src/main/scala/services/ValidationService.scala

```diff
@@ -1,11 +1,77 @@
 package com.fortysevendeg.tagless.sample
 package services
 
-import cats.data.ValidatedNel
+import cats.data.{ValidatedNec, ValidatedNel}
+import cats.effect.Async
+import cats.implicits._
 import com.fortysevendeg.tagless.sample.models._
+import org.slf4j.LoggerFactory
+
+import scala.util.matching.Regex
 
 trait ValidationService[F[_]] {
 
-  def validateUser(name: String, age: Int, gender: Char, heightAndWeight: String): F[ValidatedNel[ValidationError, UserInformation]]
+  def validateUser(
+      name: String,
+      age: Int,
+      gender: Char,
+      heightAndWeight: String): F[ValidatedNel[ValidationError, UserInformation]]
+
+}
+
+object ValidationService {
+
+  def apply[F[_]](implicit A: Async[F]): F[ValidationService[F]] =
+    A.delay(LoggerFactory.getLogger("HTTPService")).map { logger =>
+      new ValidationService[F] {
+
+        type ValidationResult[A] = ValidatedNec[ValidationError, A]
+
+        val HeightAndWeightRegEx: Regex = "(\\d+)x(\\d+)".r
+
+        def validateUser(
+            name: String,
+            age: Int,
+            gender: Char,
+            heightAndWeight: String): F[ValidatedNel[ValidationError, UserInformation]] = {
+
+          def validateAge: F[ValidationResult[Unit]] =
+            if (age >= 18) A.pure(().validNec)
+            else {
+              A.delay(logger.error(s"Error validating age of user $name. Value: $age"))
+                .as(NotAdultError.invalidNec)
+            }
+
+          def asValidatedGender(gender: Gender): ValidationResult[Gender] =
+            gender.validNec
+
+          def validateGender: F[ValidationResult[Gender]] = gender match {
+            case 'M' => asValidatedGender(Male).pure[F]
+            case 'F' => asValidatedGender(Female).pure[F]
+            case _ =>
+              A.delay(logger.error(s"Error validating gender of user $name. Value: $age"))
+                .as(InvalidGenderError.invalidNec)
+          }
+
+          def validateHeightAndWeight: F[ValidationResult[(Int, Int)]] =
+            heightAndWeight match {
+              case HeightAndWeightRegEx(height, weight) =>
```

Comment:
Oooh this is so clean! (I haven't been able to wrap my head around regexs yet!)

Reply:
I used to be a regex guy in the university, used for everything. Still applying in find and replace tasks